# Black-throated Finch Genomics
Data and code associated with population genomic study of the black-throated finch
# Recent divergence and microgeographic genetic structure in an endangered Australian songbird: the southern black-throated finch
The following is an overview of analyses and code used in Hooper et al. to examine genetic differentiation, spatial genetic structure, and demographic history in the northern black-throated finch *Poephila atropygialis* and the endangered southern black-throated finch *Poephila cincta*.

Software and programs used in this pipeline:
- [Stacks](https://catchenlab.life.illinois.edu/stacks/)
- [cutadapt](https://cutadapt.readthedocs.io/en/stable/)
- [Trimmomatic](http://www.usadellab.org/cms/?page=trimmomatic)
- [bwa](https://bio-bwa.sourceforge.net/)
- [samtools/bcftools](http://samtools.github.io/bcftools/bcftools.html)
- [STITCH](https://github.com/rwdavies/STITCH)
- [GATK](https://gatk.broadinstitute.org/hc/en-us)
- [RepeatModeler](https://github.com/Dfam-consortium/RepeatModeler)
- [vcftools](https://vcftools.sourceforge.net)
- [plink](https://www.cog-genomics.org/plink/1.9/)
- [fastStructure](https://rajanil.github.io/fastStructure/)
- [FEEMS](https://github.com/NovembreLab/feems)
- [ParseVCF_private.pl](https://github.com/ehshogren/MyzomelaPopulationGenomics/blob/main/ParseVCF_private.pl)
- [KING](https://www.kingrelatedness.com)
- [RAxML](https://github.com/stamatak/standard-RAxML)
- [vcf2phylip](https://github.com/edgardomortiz/vcf2phylip)
- [PSMC](https://github.com/lh3/psmc)
  - [vcfutils.pl](https://github.com/lh3/samtools/blob/master/bcftools/vcfutils.pl)
- [fastsimcoal27](https://cmpg.unibe.ch/software/fastsimcoal27/)
- [R](https://www.r-project.org) (v4.2.1)

## Contents

## raw read processing and aligment [WGS data]
Linked-read and short read WGS data used in this project was previously generated by McDiarmid et al. [2024](https://doi.org/10.1111/mec.17374) and Hooper et al. [2024](https://doi.org/10.1016/j.cub.2024.10.019). The raw read processing and alignment pipeline for this data is described in detail [here](https://github.com/dhooper1/Long-tailed-Finch).

## raw read processing and aligment [ddRAD data]

## variant discovery and genotyping

## population structure, genetic differentiation, and summary statistics

### principal components analyses (PCA)

### faststructure

### FEEMS

### Fst

### private alleles

## heterozygosity, relatedness, and inbreeding

### observed heterozygosity

### pairwise relatedness

### runs of homozygosity

## demographic inference

### RAxML

### PSMC

### fastsimcoal
